src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs

```diff
@@ -1053,6 +1068,190 @@ private void AddPrimarySidClaim(List<Claim> instanceClaims)
+                        claim.Properties.Add(ClaimTypes.WindowsSubAuthority, Convert.ToString(groupSid.IdentifierAuthority, CultureInfo.InvariantCulture));
+                        claim.Properties.Add(claimType, "");
+                        instanceClaims.Add(claim);
+                    }
+
+                    pSidAndAttributes = new IntPtr((long)pSidAndAttributes + Marshal.SizeOf<Interop.SID_AND_ATTRIBUTES>());
+                }
+            }
+            finally
+            {
+                safeAllocHandle.Close();
+            }
+        }
+
+        private void AddTokenClaims(List<Claim> instanceClaims, TokenInformationClass tokenInformationClass, string propertyValue)
+        {
+            // special case the anonymous identity.
+            if (_safeTokenHandle.IsInvalid)
+                return;
+
+            SafeLocalAllocHandle safeAllocHandle = SafeLocalAllocHandle.InvalidHandle;
+
+            try
+            {
+                safeAllocHandle = GetTokenInformation(_safeTokenHandle, tokenInformationClass);
+
+                Interop.CLAIM_SECURITY_ATTRIBUTES_INFORMATION claimAttributes = (Interop.CLAIM_SECURITY_ATTRIBUTES_INFORMATION)Marshal.PtrToStructure(safeAllocHandle.DangerousGetHandle(), typeof(Interop.CLAIM_SECURITY_ATTRIBUTES_INFORMATION));
+                // An attribute represents a collection of claims.  Inside each attribute a claim can be multivalued, we create a claim for each value.
+                // It is a ragged multi-dimentional array, where each cell can be of different lenghts.
+
+                // index into array of claims.
+                long offset = 0;
+
+                for (int attribute = 0; attribute < claimAttributes.AttributeCount; attribute++)
+                {
+                    IntPtr pAttribute = new IntPtr(claimAttributes.Attribute.pAttributeV1.ToInt64() + offset);
+                    Interop.CLAIM_SECURITY_ATTRIBUTE_V1 windowsClaim = (Interop.CLAIM_SECURITY_ATTRIBUTE_V1)Marshal.PtrToStructure(pAttribute, typeof(Interop.CLAIM_SECURITY_ATTRIBUTE_V1));
+
+                    // the switch was written this way, which appears to have multiple for loops, because each item in the ValueCount is of the same ValueType.  This saves the type check each item.
+                    switch (windowsClaim.ValueType)
+                    {
+                        case Interop.ClaimSecurityAttributeType.CLAIM_SECURITY_ATTRIBUTE_TYPE_STRING:
+                            IntPtr[] stringPointers = new IntPtr[windowsClaim.ValueCount];
+                            Marshal.Copy(windowsClaim.Values.ppString, stringPointers, 0, (int)windowsClaim.ValueCount);
+
+                            for (int item = 0; item < windowsClaim.ValueCount; item++)
+                            {
+                                Claim c = new Claim(windowsClaim.Name, Marshal.PtrToStringAuto(stringPointers[item]), ClaimValueTypes.String, _issuerName, _issuerName, this);
```

Comment:
Based on some of the earlier marshaling code, I'm assuming we're not particularly concerned with perf in this stuff. If we were, though, it'd be better to avoid the IntPtr[] and just get each IntPtr here when it's needed to get the string.